File: apps/mobile/test/recipe_activity_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:pi_mob/src/domain/mobile_state.dart';
import 'package:pi_mob/src/recipe_activity/domain/recipe_activity.dart';
import 'package:pi_mob/src/recipe_activity/widgets/recipe_activity_view.dart';

StreamEventState event(int cursor, String type, Map<String, Object?> payload) =>
    StreamEventState(
      hostId: 'host',
      streamId: 'session:s',
      cursor: StreamCursor.parse('$cursor'),
      eventId: 'event-$cursor',
      type: type,
      payload: payload,
      occurredAt: DateTime.utc(2026, 1, 1, 0, 0, cursor),
    );

void main() {
  test('projects normalized activities without promoting reasoning text', () {
    final result = projectRecipeActivities([
      event(1, 'turn.started', {'sessionId': 's', 'turnId': 't'}),
      event(2, 'reasoning.started', {
        'sessionId': 's',
        'turnId': 't',
        'contentBlockId': 'r',
      }),
      event(3, 'reasoning.delta', {
        'sessionId': 's',
        'turnId': 't',
        'contentBlockId': 'r',
        'text': 'private',
      }),
      event(4, 'reasoning.completed', {
        'sessionId': 's',
        'turnId': 't',
        'contentBlockId': 'r',
      }),
      event(5, 'tool.started', {
        'sessionId': 's',
        'turnId': 't',
        'toolCallId': 'a',
        'toolName': 'read',
        'arguments': {'path': 'README.md'},
      }),
      event(6, 'tool.completed', {
        'sessionId': 's',
        'turnId': 't',
        'toolCallId': 'a',
        'result': {'output': 'ok'},
      }),
    ]);
    expect(result, hasLength(2));
    expect(result.first.kind, RecipeActivityKind.thinking);
    expect(result.first.status, RecipeActivityStatus.completed);
    expect(result.first.title, 'Thinking');
    expect(result.last.toolName, 'read');
    expect(result.last.status, RecipeActivityStatus.completed);
  });

  test('replay dedupe and bounds arguments', () {
    final start = event(1, 'tool.started', {
      'sessionId': 's',
      'turnId': 't',
      'toolCallId': 'a',
      'toolName': 'bash',
      'arguments': 'x' * 400,
    });
    final result = projectRecipeActivities([
      start,
      start,
      event(2, 'reasoning.started', {
        'sessionId': 's',
        'turnId': 't',
        'contentBlockId': 'a',
      }),
    ]);
    expect(result, hasLength(1));
    expect(result.single.kind, RecipeActivityKind.tool);
    expect(result.single.arguments, hasLength(240));
    expect(result.single.truncation, isNull);
  });

  test('rejects oversized and malformed activity identities', () {
    final result = projectRecipeActivities([
      event(1, 'tool.started', {
        'sessionId': 's',
        'turnId': 't',
        'toolCallId': 'a' * 129,
        'toolName': 'bash',
      }),
      event(2, 'tool.started', {
        'sessionId': 's',
        'turnId': 't',
        'toolCallId': 'valid',
      }),
    ]);
    expect(result, isEmpty);
  });

  test('preserves cancelled status without fabricated truncation', () {
    final result = projectRecipeActivities([
      event(1, 'tool.started', {
        'sessionId': 's',
        'turnId': 't',
        'toolCallId': 'a',
        'toolName': 'bash',
      }),
      event(2, 'tool.cancelled', {
        'sessionId': 's',
        'turnId': 't',
        'toolCallId': 'a',
      }),
    ]);
    expect(result.single.status, RecipeActivityStatus.cancelled);
    expect(result.single.truncation, isNull);
  });

  testWidgets('recipe activity is collapsed and expands details', (
    tester,
  ) async {
    final activity = RecipeActivity(
      kind: RecipeActivityKind.tool,
      sessionId: 's',
      turnId: 't',
      activityId: 'a',
      ordinal: 0,
      status: RecipeActivityStatus.completed,
      timing: RecipeTiming(startedAt: DateTime.utc(2026)),
      title: 'read',
      toolName: 'read',
      arguments: '{}',
      output: 'ok',
    );
    await tester.pumpWidget(
      MaterialApp(home: RecipeActivityView(activity: activity)),
    );
    expect(find.text('Arguments: {}'), findsNothing);
    await tester.tap(find.byType(InkWell));
    await tester.pumpAndSettle();
    expect(find.text('Arguments: {}'), findsOneWidget);
  });
}

File: packages/bridge/test/recipe-activity.test.ts
import { describe, expect, test } from "bun:test";
import { projectRecipeActivities } from "../src/pi/recipe-activity";

const sessionId = "11111111-1111-1111-1111-111111111111";
const event = (type: string, payload: Record<string, unknown>, occurredAt: string) => ({
  type,
  payload: { sessionId, ...payload },
  occurredAt,
} as const);

describe("recipe activity projector", () => {
  test("projects sparse reasoning and tool lifecycles with stable ordinals", () => {
    const events = [
      event("turn.started", { turnId: "turn-1" }, "2026-01-01T00:00:00.000Z"),
      event("reasoning.started", { turnId: "turn-1", contentBlockId: "reason-1" }, "2026-01-01T00:00:01.000Z"),
      event("reasoning.delta", { turnId: "turn-1", contentBlockId: "reason-1", text: "private reasoning" }, "2026-01-01T00:00:02.000Z"),
      event("reasoning.completed", { turnId: "turn-1", contentBlockId: "reason-1" }, "2026-01-01T00:00:03.000Z"),
      event("tool.started", { turnId: "turn-1", toolCallId: "tool-1", toolName: "read", arguments: { path: "README.md" } }, "2026-01-01T00:00:04.000Z"),
      event("tool.output", { turnId: "turn-1", toolCallId: "tool-1", output: "ok", retainedBytes: 2, totalBytes: 2, isTruncated: false }, "2026-01-01T00:00:05.000Z"),
      event("tool.completed", { turnId: "turn-1", toolCallId: "tool-1", result: { output: "ok" } }, "2026-01-01T00:00:06.000Z"),
    ];
    const result = projectRecipeActivities(events);
    expect(result).toHaveLength(2);
    expect(result.map((item) => item.ordinal)).toEqual([0, 1]);
    expect(result[0]).toMatchObject({ kind: "thinking", status: "completed", title: "Thinking" });
    expect(result[0]).not.toHaveProperty("providerSummary");
    expect(result[1]).toMatchObject({ kind: "tool", status: "completed", toolName: "read", output: '{"output":"ok"}' });
    expect(result[1]!.timing.durationMs).toBe(2000);
  });

  test("rejects oversized ids before identity and preserves the first kind on collisions", () => {
    const oversized = "x".repeat(129);
    const result = projectRecipeActivities([
      event("tool.started", { turnId: "t", toolCallId: oversized, toolName: "bash" }, "2026-01-01T00:00:00.000Z"),
      event("tool.started", { turnId: "t", toolCallId: "a", toolName: "bash" }, "2026-01-01T00:00:01.000Z"),
      event("reasoning.started", { turnId: "t", contentBlockId: "a" }, "2026-01-01T00:00:02.000Z"),
    ]);
    expect(result).toHaveLength(1);
    expect(result[0]).toMatchObject({ kind: "tool", activityId: "a" });
  });

  test("does not fabricate tool fields for malformed or absent values", () => {
    const circular: Record<string, unknown> = {};
    circular.self = circular;
    const result = projectRecipeActivities([
      event("tool.started", { turnId: "t", toolCallId: "a" }, "2026-01-01T00:00:00.000Z"),
      event("tool.started", { turnId: "t", toolCallId: "b", toolName: "bash", arguments: circular }, "2026-01-01T00:00:01.000Z"),
      event("tool.completed", { turnId: "t", toolCallId: "b" }, "2026-01-01T00:00:02.000Z"),
    ]);
    expect(result).toHaveLength(1);
    expect(result[0]).toMatchObject({ activityId: "b", toolName: "bash" });
    expect(result[0]).not.toHaveProperty("arguments");
    expect(result[0]).not.toHaveProperty("output");
    expect(result[0]).not.toHaveProperty("truncation");
  });

  test("only includes truncation when source metadata exists", () => {
    const result = projectRecipeActivities([
      event("tool.started", { turnId: "t", toolCallId: "a", toolName: "bash", arguments: "x".repeat(400) }, "2026-01-01T00:00:00.000Z"),
      event("tool.completed", { turnId: "t", toolCallId: "a", result: "y".repeat(400) }, "2026-01-01T00:00:01.000Z"),
    ]);
    expect(result[0]!.arguments).toHaveLength(240);
    expect(result[0]!.output).toHaveLength(240);
    expect(result[0]).not.toHaveProperty("truncation");
  });

  test("tracks failed and cancelled terminal states", () => {
    const failed = projectRecipeActivities([
      event("tool.started", { turnId: "t", toolCallId: "a", toolName: "bash" }, "2026-01-01T00:00:00.000Z"),
      event("tool.failed", { turnId: "t", toolCallId: "a", output: "failure", errorInfo: { code: "internal_error", message: "failure", retryable: false }, retainedBytes: 3, totalBytes: 7, isTruncated: true }, "2026-01-01T00:00:01.000Z"),
    ]);
    expect(failed[0]).toMatchObject({ status: "failed", errorInfo: { code: "internal_error", retryable: false }, truncation: { retainedBytes: 3, totalBytes: 7, isTruncated: true } });

    const cancelled = projectRecipeActivities([
      event("tool.started", { turnId: "t", toolCallId: "b", toolName: "bash" }, "2026-01-01T00:00:00.000Z"),
      event("tool.cancelled", { turnId: "t", toolCallId: "b" }, "2026-01-01T00:00:01.000Z"),
    ]);
    expect(cancelled[0]).toMatchObject({ status: "cancelled" });
  });
});
